tests/drivers/random/randomnumgen_basic_api/src/main.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <random.h>
+#include <ztest.h>
+
+/*
+ * @addtogroup t_randomnumgen_basic_api
+ * @{
+ * @defgroup t_random_get_entropy test_random_get_entropy
+ * @brief TestPurpose: verify Random number generator works
+ * @details
+ * - Test Steps
+ *   -# Read random numbers from RNG driver.
+ *   -# Verify whether buffer overflow occured or not.
+ *   -# Verify whether buffer completly filled or not.
+ * - Expected Results
+ *   -# Random number should be generated.
+ * @}
+ */
+
+#define BUFFER_LENGTH 10
+#define RECHECK_RANDOM_ENTROPY 0x10
+
+static int random_entropy(struct device *dev, char *buffer, char num)
+{
+	int ret, i;
+	int count = 0;
+
+	memset(buffer, num, BUFFER_LENGTH);
+
+	/* The BUFFER_LENGTH-1 is used so the driver will not
+	 * write the last byte of the buffer. If that last
+	 * byte is not 0 on return it means the driver wrote
+	 * outside the passed buffer, and that should never
+	 * happen.
+	 */
+	ret = random_get_entropy(dev, buffer, BUFFER_LENGTH-1);
+	if (ret) {
+		TC_PRINT("Error: random_get_entropy failed: %d\n", ret);
+		return TC_FAIL;
+	}
+	if (buffer[BUFFER_LENGTH-1] != num) {
+		TC_PRINT("Error: random_get_entropy buffer overflow\n");
+		return TC_FAIL;
+	}
+
+	for (i = 0; i < BUFFER_LENGTH-1; i++) {
+		TC_PRINT("  0x%02x\n", buffer[i]);
+		if (buffer[i] == num)
+			count++;
+	}
+
+	if (count >= 2)
+		return RECHECK_RANDOM_ENTROPY;
+	else
+		return TC_PASS;
+}
+
+/*
+ * Function invokes the get_entropy callback in driver
+ * to get the random data and fill to passed buffer
+ */
+static int get_entropy(void)
+{
+	struct device *dev;
+	u8_t buffer[BUFFER_LENGTH] = {0};
+	int ret;
+
+	dev = device_get_binding(CONFIG_RANDOM_NAME);
+	if (!dev) {
+		TC_PRINT("error: no random device\n");
+		return TC_FAIL;
+	}
+
+	TC_PRINT("random device is %p, name is %s\n",
+			dev, dev->config->name);
+
+	ret = random_entropy(dev, buffer, 0);
+
+	/* Check whether 20% or more of buffer still filled with default
+	 * value(0), if yes then recheck again by filling nonzero value(0xa5)
+	 * to buffer. Recheck random_entropy and verify whether 20% or more
+	 * of buffer filled with value(0xa5) or not.
+	 */
+	if (ret == RECHECK_RANDOM_ENTROPY) {
```

Comment:
It is unclear to me what we are testing for. 20% seems arbitrary, and the chance of this just failing will depend on the size of the test. With 10 bytes given, each byte has a 1/256 chance of being the pattern value, which I think gives us a probability of 1.34e-08 for a false positive here. I'm not sure rechecking is really necessary (it just squares the probability).

Reply:
This isn't testing the quality of the random number here -- it's too simplistic of a test to do this.  The reasoning behind the test -- although it's far from being perfect -- is to see if each driver's implementation of `random_get_entropy()` is in fact filling up the whole buffer.  That's why it tests twice; the possibility the random generator indeed generating more than 20% of 0 bytes twice is pratically none.

Reply:
about 1.34e-08 for one pass, and the square of that for two passes. You probably won't ever see it have to do two passes.